src/meta/processors/job/BalancePlan.cpp

```diff
@@ -92,36 +100,48 @@ void BalancePlan::invoke() {
           stopped = stopped_;
         }
         if (finished) {
-          CHECK_EQ(j, this->buckets_[i].size() - 1);
-          onFinished_();
-        } else if (j + 1 < this->buckets_[i].size()) {
-          auto& task = this->tasks_[this->buckets_[i][j + 1]];
+          CHECK_EQ(j, buckets_[i].size() - 1);
+          onFinished_(stopped ? meta::cpp2::JobStatus::STOPPED : meta::cpp2::JobStatus::FAILED);
+        } else if (j + 1 < buckets_[i].size()) {
+          auto& task = tasks_[buckets_[i][j + 1]];
           if (tasks_[taskIndex].spaceId_ == task.spaceId_ &&
               tasks_[taskIndex].partId_ == task.partId_) {
             LOG(INFO) << "Skip the task for the same partId " << task.partId_;
             task.ret_ = BalanceTaskResult::FAILED;
           }
-          if (stopped) {
-            task.ret_ = BalanceTaskResult::INVALID;
-          }
+          task.ret_ = BalanceTaskResult::INVALID;
           task.invoke();
+        } else {
+          size_t index = curIndex_.fetch_add(1, std::memory_order_relaxed);
```

Comment:
`curIndex_` should not use `memory_order_relaxed` in this PR. You could just reuse the lock here.